lib/plug/adapters/cowboy2.ex

```diff
@@ -0,0 +1,306 @@
+      _ ->
+        cowboy_options
+    end
+  end
+
+  defp normalize_cowboy_options(cowboy_options, :http) do
+    Keyword.put_new cowboy_options, :port, 4000
+  end
+
+  defp normalize_cowboy_options(cowboy_options, :https) do
+    assert_ssl_options(cowboy_options)
+    cowboy_options = Keyword.put_new cowboy_options, :port, 4040
+    cowboy_options = Enum.reduce [:keyfile, :certfile, :cacertfile, :dhfile], cowboy_options, &normalize_ssl_file(&1, &2)
+    cowboy_options = Enum.reduce [:password], cowboy_options, &to_charlist(&2, &1)
+    cowboy_options
+  end
+
+  defp to_args(opts, non_keyword_opts) do
+    opts = Keyword.delete(opts, :otp_app)
+    {ref, opts} = Keyword.pop(opts, :ref)
+    {dispatch, opts} = Keyword.pop(opts, :dispatch)
+    {num_acceptors, opts} = Keyword.pop(opts, :acceptors, 100)
+    {protocol_options, opts} = Keyword.pop(opts, :protocol_options, [])
+
+    dispatch = :cowboy_router.compile(dispatch)
+    {extra_options, transport_options} = Keyword.split(opts, @protocol_options)
+
+    extra_options = Keyword.put_new(extra_options, :stream_handlers, [Plug.Adapters.Cowboy2.BadResponseCheck, :cowboy_stream_h])
+
+    protocol_options = %{
+      env: %{
+        dispatch: dispatch
+      }
+    }
+    |> Map.merge(:maps.from_list(protocol_options ++ extra_options))
+
+    transport_options = Keyword.put_new(transport_options, :num_acceptors, num_acceptors)
+
+    [ref, non_keyword_opts ++ transport_options, protocol_options]
+  end
+
+  defp build_ref(plug, scheme) do
+    Module.concat(plug, scheme |> to_string |> String.upcase)
+  end
+
+  defp dispatch_for(plug, opts) do
+    opts = plug.init(opts)
+    [{:_, [{:_, Plug.Adapters.Cowboy2.Handler, {plug, opts}}]}]
+  end
+
+  defp normalize_ssl_file(key, cowboy_options) do
+    value = cowboy_options[key]
+
+    cond do
+      is_nil(value) ->
+        cowboy_options
+      Path.type(value) == :absolute ->
+        put_ssl_file cowboy_options, key, value
+      true ->
+        put_ssl_file cowboy_options, key, Path.expand(value, otp_app(cowboy_options))
+    end
+  end
+
+  defp assert_ssl_options(cowboy_options) do
+    unless Keyword.has_key?(cowboy_options, :key) or
+           Keyword.has_key?(cowboy_options, :keyfile) do
+      fail "missing option :key/:keyfile"
+    end
+    unless Keyword.has_key?(cowboy_options, :cert) or
+           Keyword.has_key?(cowboy_options, :certfile) do
+      fail "missing option :cert/:certfile"
+    end
+  end
+
+  defp put_ssl_file(cowboy_options, key, value) do
+    value = to_charlist(value)
+    unless File.exists?(value) do
+      fail "the file #{value} required by SSL's #{inspect key} either does not exist, or the application does not have permission to access it"
+    end
+    Keyword.put(cowboy_options, key, value)
+  end
+
+  defp otp_app(cowboy_options) do
+    if app = cowboy_options[:otp_app] do
+      Application.app_dir(app)
```

Comment:
Probably this is not optimal way to fech priv_dir path, see here: https://elixirforum.com/t/elixir-blog-posts/150/185.

Reply:
It is fine for our use case though. :)